### ISSUE TITLE
ATRIndicator

### ISSUE THREAD

**Issue by JamesKBowler** [mhallsmoore/qstrader#222]
Hi all,

Made this today, thought I would share. Anyone has a faster alternative please let me know.

```
from collections import deque
import numpy as np

from qstrader.indicators.base import AbstractIndicator
from qstrader.event import EventType

class ATRIndicator(AbstractIndicator):
    """
    Requires:
    ticker - The ticker symbol being used for moving averages
    window - Lookback period for exponential moving average
    true range window - Period for calculating ATR
    
    # Some information about TR
    http://www.investopedia.com/terms/a/atr.asp
    ch_cl = current high - current low
    ch_pc = current high - previous close (absolute value)
    cl_pc = current low - previous close (absolute value)
    tr = max(ch_cl, ch_pc, cl_pc)
    """
    def __init__(self, ticker, window=90):
        self.ticker = ticker
        self.w_bars = deque(maxlen=self.window)
        self.tr_bars = deque(maxlen=self.window)
        self.atr_values = deque(maxlen=self.window)
        self.bars = 0
        self.atr = 0

    def calculate(self, event):
        if (
            event.type == EventType.BAR and event.ticker == self.ticker
        ):
            self.bars += 1
            # Add latest adjusted closing price to the window of bars
            self.w_bars.append(event.adj_close_price)
            # Calculate True Range
            ch_cl = event.high_price - event.low_price
            ch_pc = 0.0
            cl_pc = 0.0
            if self.bars != 1:
                ch_pc = abs(event.high_price - self.w_bars[-2])
                cl_pc = abs(event.low_price - self.w_bars[-2])
            tr = max(ch_cl, ch_pc, cl_pc)
            self.tr_bars.append(tr)
            # Calulate first average point
            if self.bars == self.window:
                avg = np.mean(self.tr_bars)
                self.atr_values.append(avg)
            # Calulate exponential moving average
            elif self.bars > self.window:
                self.atr = int(
                    (self.atr_values[-1] * (self.window-1) + self.tr_bars[-1]
                        ) // self.window
                )
                self.atr_values.append(self.atr)
```

Access the latest ATR value by calling:

`Indicators.dict_ind['ATRIndicator']['APPL'].atr`

**Comment by JamesKBowler** [comment]
Here is the abstract class.
```
from abc import ABCMeta, abstractmethod
from collections import defaultdict


class AbstractIndicator(object):
    """
    AbstractIndicator is an abstract base class providing an
    interface for all subsequent (inherited) indicator handling
    objects.

    The goal of a (derived) Indicator object is to generate indicator
    values for particular symbols based on the inputs of bars
    generated from a PriceHandler (derived) object.

    This is designed to work both with historic and live data as
    the Indicator object is agnostic to data location.
    """

    __metaclass__ = ABCMeta

    @abstractmethod
    def calculate(self, event):
        """
        Provides the mechanisms to calculate the list of indicators.
        """
        raise NotImplementedError("Should implement calculate()")


class Indicators(AbstractIndicator):
    """
    Indicators is a collection of Indicator
    """
    def __init__(self, *indicators):
        self._dict_ind = defaultdict(dict)
        for i in indicators:
            self._dict_ind[i.__class__.__name__][i.ticker] = i

    def calculate(self, event):
        for indicator, obj in self._dict_ind.items():
            try:
                obj[event.ticker].calculate(event)
            except KeyError:
                pass
```

**Comment by enriqueromualdez** [comment]
Hi @JamesKBowler,

Went through the code and it looks perfectly fine to me. 

I haven't made an separate ATR indicator, but I've coded up an ATR Stop Loss and Keltner Channels which use the ATR. We have very similar code minus the use of the AbstractIndicator class. 

I've noticed that the accuracy of the ATR is off for the first year's worth of bar data (252 bars). This was when compared to TradingView's ATR. I incorporated a lookback in the StrategyObject to allow the indicators to match up with TradingView's ATR before calculating any entry signals.
